feat: add topics model to database schema

wanlingt opened the pull request:
## Problem

The current database schema does not allow for posts to be organised according to agency-specific topics (formerly referred to as 'categories')

Closes #396 

## Solution

- Add a new model `Topic`
- Modify the `Post` model such that posts can be organised into topics. `topicId` has been added as a FK

Details ...
- Each topic knows its immediate parent; child topics carry a foreign key (`parentId`) to their parent topic. For parent topics, the `parentId` will be null.
- In the `posts` table, `topicId` is nullable. This means that `topics` are optional for each post.

## Tests
- After running the migration script, verify that the ERD diagram is updated with the new `Topic` model


## Deploy Notes

Create new `topics` table and update the `posts` table before deploying by running the new migration scripts.

**New scripts**:

- `20211010101959-create-topics.js` : create topics table
- `20211011014741-add-topic-to-posts.js` : add topicId as a new column in posts table

wanlingt committed:
feat(categories): add categories model

add categories as a new model, modify posts model to include categoryId as a FK

#396

wanlingt committed:
feat(categories): add categories model

add categories as a new model, modify posts model to include categoryId as a FK

wanlingt committed:
feat(sequelize): create migration scripts for categories table

create migration script for categories table, add new column referencing categoryId to posts table

#396

wanlingt committed:
feat(sequelize): create migration scripts for categories table

create migration script for categories table, add new column referencing categoryId to posts table

wanlingt committed:
feat(sequelize): add seed data for categories

re #396

wanlingt committed:
test: update jest-db with category model

re #396

wanlingt committed:
docs(readme): update readme to reflect new tables

changes made:
- new category table
- posts table includes categoryId as a FK

nauynix commented on server/migrations/20210922101959-create-categories.js:
```diff
@@ -0,0 +1,49 @@
+          key: 'id',
+        },
+        onUpdate: 'CASCADE',
+        onDelete: 'SET NULL', //Cascade?
```
Have you resolved if this should be set null or cascade?

LoneRifle replied:
should be `SET NULL`. I don't think it's a good idea if the parent gets deleted, leaving orphans in its wake.

nauynix commented on server/migrations/20210923014741-add-category-to-posts.js:
```diff
@@ -0,0 +1,20 @@
+  },
+
+  down: async (queryInterface) => {
+    await queryInterface.removeColumn('posts', 'agencyId')
```
is `agencyId` supposed to be `categoryId` instead?

wanlingt replied:
yes my bad

nauynix commented on shared/src/types/base/category.ts:
```diff
@@ -0,0 +1,10 @@
+
+export const Category = BaseModel.extend({
+  catname: z.string(),
+  parentId: z.string(),
```
Is the parentId supposed to be a number?

wanlingt replied:
yup thanks for spotting!

LoneRifle commented on server/migrations/20210922101959-create-categories.js:
```diff
@@ -0,0 +1,49 @@
+        primaryKey: true,
+        type: Sequelize.INTEGER,
+      },
+      catname: {
```
```suggestion
      name: {
```
and elsewhere

LoneRifle commented on server/seeders/20210928063247-was-add-categories.js:
```diff
@@ -0,0 +1,57 @@
+'use strict'
```
I think let's not worry about having a seeder script for this for now. Having chatted with others and seeing how AskGov has been most recently used, I have a sense that we should make _both_ categories _and_ tags optional. This would grant the agency some flexibility over how their information should be organised. They could either:
- Use categories only to organise information in traditional manner;
- Use tags only for a more ad-hoc approach that allows them to group questions under multiple (possibly cross-agency) topics, or;
- Use both approaches.

wanlingt replied:
removed seeder script

LoneRifle commented on server/src/models/categories.model.ts:
```diff
@@ -0,0 +1,26 @@
+  sequelize: Sequelize,
+  { Agency }: { Agency: ModelCtor<Agency> },
+): ModelCtor<Category> => {
+  const Category: ModelCtor<Category> = sequelize.define('category', {
```
Consider typing it as `ModelDef<Category>` (the shared type you have defined). This allows you to avoid having to explicitly define the Category in line 5.

LoneRifle commented on shared/src/types/base/category.ts:
```diff
@@ -0,0 +1,10 @@
+import { BaseModel } from './common'
+
+export const Category = BaseModel.extend({
+  catname: z.string(),
```
```suggestion
  name: z.string(),
```

LoneRifle commented on server/src/models/categories.model.ts:
```diff
@@ -0,0 +1,26 @@
+  { Agency }: { Agency: ModelCtor<Agency> },
+): ModelCtor<Category> => {
+  const Category: ModelCtor<Category> = sequelize.define('category', {
+    catname: {
```
```suggestion
    name: {
```

LoneRifle commented on server/migrations/20210922101959-create-categories.js:
```diff
@@ -0,0 +1,49 @@
+        type: Sequelize.DATE,
+      },
+      agencyId: {
+        allowNull: true, //to change to false once we implement this across all agencies
```
```suggestion
        allowNull: false
```
I think this should be false from the get-go, you can't really have categories without an agency. An agency however can get away without having categories, particularly since posts already have a nullable categoryId.

LoneRifle reviewed (changes requested):
I realise I'm setting a poor example by reviewing this PR at this hour. Nonetheless, do take a look at my initial comments.

LoneRifle commented on server/migrations/20210922101959-create-categories.js:
```diff
@@ -0,0 +1,49 @@
+        type: Sequelize.TEXT,
+        allowNull: false,
+      },
+      createdAt: {
```
We probably need a nullable description here.

LoneRifle commented:
actually, shall we rename this to topics instead, given that's what the public knows them by?

wanlingt commented:
> actually, shall we rename this to topics instead, given that's what the public knows them by?

can

wanlingt committed:
refactor(topics): rename categories to topics

- rename categories to topics
- add description to topics model
- convert ModelCtor to
ModelDefined in topics and posts model

re #396

wanlingt committed:
update README with new ER diagram

wanlingt committed:
fix(topic-post-migration): remove outdated comment in add-topic-to-posts migration script

re #396

wanlingt committed:
fix(posts): revert back to using ModelCtor for posts model
